lib/Conversion/VerifToSMT/VerifToSMT.cpp

```diff
@@ -160,21 +378,70 @@ struct ConvertVerifToSMTPass
 } // namespace
 
 void circt::populateVerifToSMTConversionPatterns(TypeConverter &converter,
-                                                 RewritePatternSet &patterns) {
-  patterns.add<VerifAssertOpConversion, LogicEquivalenceCheckingOpConversion>(
-      converter, patterns.getContext());
+                                                 RewritePatternSet &patterns,
+                                                 Namespace &names) {
+  patterns.add<VerifAssertOpConversion, VerifAssumeOpConversion,
+               LogicEquivalenceCheckingOpConversion>(converter,
+                                                     patterns.getContext());
+  patterns.add<VerifBoundedModelCheckingOpConversion>(
+      converter, patterns.getContext(), names);
 }
 
 void ConvertVerifToSMTPass::runOnOperation() {
   ConversionTarget target(getContext());
   target.addIllegalDialect<verif::VerifDialect>();
-  target.addLegalDialect<smt::SMTDialect, arith::ArithDialect>();
+  target.addLegalDialect<smt::SMTDialect, arith::ArithDialect, scf::SCFDialect,
+                         func::FuncDialect>();
   target.addLegalOp<UnrealizedConversionCastOp>();
 
+  // Check BMC ops contain only one assertion (done outside pattern to avoid
+  // issues with whether assertions are/aren't lowered yet)
+  SymbolTable symbolTable(getOperation());
+  getOperation().walk(
+      [&](Operation *op) { // Check there is exactly one assertion
+        if (isa<verif::BoundedModelCheckingOp>(op)) {
+          SmallVector<mlir::Operation *> worklist;
+          int numAssertions = 0;
+          op->walk([&](Operation *curOp) {
+            if (isa<verif::AssertOp>(curOp))
+              numAssertions++;
+            if (auto inst = dyn_cast<InstanceOp>(curOp))
+              worklist.push_back(symbolTable.lookup(inst.getModuleName()));
+          });
+          // TODO: probably negligible compared to actual model checking time
+          // but cacheing the assertion count of modules would speed this up
+          while (!worklist.empty()) {
+            auto *module = worklist.pop_back_val();
+            module->walk([&](Operation *curOp) {
+              if (isa<verif::AssertOp>(curOp))
+                numAssertions++;
+              if (auto inst = dyn_cast<InstanceOp>(curOp))
+                worklist.push_back(symbolTable.lookup(inst.getModuleName()));
+            });
+            if (numAssertions > 1)
+              break;
+          }
+          if (numAssertions > 1) {
+            op->emitError(
+                "bounded model checking problems with multiple assertions are "
+                "not yet "
+                "correctly handled - instead, you can assert the "
+                "conjunction of your assertions");
+            signalPassFailure();
```

Comment:
We could early exit the pass here. Otherwise fine to merge from my side. Thanks for moving it out!